In [1]:
#importing pandas
import pandas as pd

In [2]:
#reading the dataset
dataset=pd.read_csv("music_streaming.csv")

In [3]:
#examining the dataset
dataset.head()

Artist Name                             Track Name  Popularity  \
0             Bruno Mars  That's What I Like (feat. Gucci Mane)        60.0   
1                 Boston                           Hitch a Ride        54.0   
2          The Raincoats                     No Side to Fall In        35.0   
3                   Deno             Lingo (feat. J.I & Chunkz)        66.0   
4  Red Hot Chili Peppers      Nobody Weird Like Me - Remastered        53.0   

   danceability  energy   key  loudness  mode  speechiness  acousticness  \
0         0.854   0.564   1.0    -4.964     1       0.0485      0.017100   
1         0.382   0.814   3.0    -7.230     1       0.0406      0.001100   
2         0.434   0.614   6.0    -8.334     1       0.0525      0.486000   
3         0.853   0.597  10.0    -6.528     0       0.0555      0.021200   
4         0.167   0.975   2.0    -4.279     1       0.2160      0.000169   

   instrumentalness  liveness  valence    tempo  duration_in min/ms  \
0               NaN    0.0849   0.8990  134.071            234596.0   
1          0.004010    0.1010   0.5690  116.454            251733.0   
2          0.000196    0.3940   0.7870  147.681            109667.0   
3               NaN    0.1220   0.5690  107.033            173968.0   
4          0.016100    0.1720   0.0918  199.060            229960.0   

   time_signature  Genre  
0               4      5  
1               4     10  
2               4      6  
3               4      5  
4               4     10

In [7]:
#see if the dataset has null values
print(dataset.isnull().sum())

Artist Name              0
Track Name               0
Popularity             394
danceability             0
energy                   0
key                   1743
loudness                 0
mode                     0
speechiness              0
acousticness             0
instrumentalness      3587
liveness                 0
valence                  0
tempo                    0
duration_in min/ms       0
time_signature           0
Genre                    0
dtype: int64


In [8]:
#imputating the null values
dataset['Popularity'].fillna(dataset['Popularity'].median(), inplace=True)
dataset['key'].fillna(dataset['key'].mode()[0], inplace=True)
dataset['instrumentalness'].fillna(0, inplace=True)

In [9]:
#no more null values :)
print(dataset.isnull().sum())

Artist Name           0
Track Name            0
Popularity            0
danceability          0
energy                0
key                   0
loudness              0
mode                  0
speechiness           0
acousticness          0
instrumentalness      0
liveness              0
valence               0
tempo                 0
duration_in min/ms    0
time_signature        0
Genre                 0
dtype: int64


In [10]:
#no duplicates
print(dataset.duplicated().sum())

0


In [11]:
#exmaining datatypes of each column
print(dataset.dtypes)

Artist Name            object
Track Name             object
Popularity            float64
danceability          float64
energy                float64
key                   float64
loudness              float64
mode                    int64
speechiness           float64
acousticness          float64
instrumentalness      float64
liveness              float64
valence               float64
tempo                 float64
duration_in min/ms    float64
time_signature          int64
Genre                   int64
dtype: object


In [12]:
#removing duplices of artist name + track name (so the quries will be fair)
dataset.drop_duplicates(subset=['Artist Name', 'Track Name'], inplace=True)

In [ ]:
#nader cleaning
df.isna().sum()/len(df)*100
mean_value = df['Popularity'].mean()
df['Popularity'].fillna(mean_value, inplace=True)
#df.dropna(subset=['Popularity'], inplace=True)
df = df.drop_duplicates()
# Calculate the mode of 'key' column within each group defined by 'Genre'
mode_per_genre = df.groupby('Genre')['key'].apply(lambda x: x.mode().iloc[0])

# Fill missing values in 'key' column with the mode of their respective genre
df['key'] = df.apply(lambda row: mode_per_genre[row['Genre']] if pd.isna(row['key']) else row['key'], axis=1)
imputerProMax = IterativeImputer()
impute_columns = df.drop(["Artist Name", "Track Name"], axis=1).columns
imputerProMax.fit(df[impute_columns])
imputed_values = imputerProMax.transform(df[impute_columns])
df[impute_columns] = imputed_values
df.isna().sum()/len(df)*100
def convert_to_milliseconds(duration):
    if duration < 60:
        return duration * 60000
    else:
        return duration

df['duration_in min/ms'] = df['duration_in min/ms'].apply(convert_to_milliseconds)

print(df[df.duplicated()])
df = df.drop_duplicates()
def min_max_normalize(column):
    return (column - column.min()) / (column.max() - column.min())

df['tempo'] = min_max_normalize(df['tempo'])
df.head(100)
df.info()

In [4]:
#installing libraries
!pip install pyspark
!pip install findspark
!pip install pyarrow==1.0.0

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 317.0/317.0 MB 3.5 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for pyspark: filename=pyspark-3.5.1-py2.py3-none-any.whl size=317488491 sha256=6b6a2db2098bc0ec5a6aff6ff298cc95c12d12c2a01496a9c1562f250438d416
  Stored in directory: /root/.cache/pip/wheels/80/1d/60/2c256ed38dddce2fdd93be545214a63e02fbd8d74fb0b7f3a6
Successfully built pyspark
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 9.1 MB/s eta 0:00:00
  error: subprocess-exited-with-error
  
  × pip subprocess to install build dependencies did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  Installing build dependencies ... error
error: subprocess-exited-with-error

× pip subprocess to install build dependencies did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely

# SQL

In [13]:
#improting libraries
import findspark
findspark.init()#adding spark installiation to our python path

In [14]:
#improting libraries
from pyspark import SparkContext, SparkConf
from pyspark.sql import SparkSession

In [15]:
#creating our app, dataframe and temporary view/table
spark = SparkSession.builder \
    .appName("MusicStreamingAnalysis") \
    .getOrCreate()

spark_df = spark.createDataFrame(dataset)
spark_df.createOrReplaceTempView("songs")

a)Which genre has the highest average popularity?

In [16]:
query = """
        SELECT Genre, AVG(Popularity) AS avg_popularity
        FROM songs
        GROUP BY Genre
        ORDER BY avg_popularity DESC
        LIMIT 1
        """
result = spark.sql(query)

result.show()

+-----+-----------------+
|Genre|   avg_popularity|
+-----+-----------------+
|    4|56.70161290322581|
+-----+-----------------+



b)Display which artists have recorded the most number of songs with a duration of more than 5 minutes

In [17]:
query = """
        SELECT `Artist Name`, COUNT(*) AS num_songs
        FROM songs
        WHERE `duration_in min/ms` > 5*60000
        GROUP BY `Artist Name`
        ORDER BY num_songs DESC
        """

result = spark.sql(query)

result.show()

+--------------------+---------+
|         Artist Name|num_songs|
+--------------------+---------+
|           Metallica|       19|
|                TOOL|       11|
|        Led Zeppelin|       11|
|         Arcade Fire|        8|
|           Pearl Jam|        8|
|           Aerosmith|        8|
|       Joe Bonamassa|        6|
|        Dire Straits|        5|
|                  U2|        5|
|   Avenged Sevenfold|        5|
|Kenny Wayne Shepherd|        5|
|             Pantera|        5|
|               Opeth|        5|
|     The Stone Roses|        4|
|                 Yes|        4|
|      Phani Narayana|        4|
|       The Beta Band|        4|
|Creedence Clearwa...|        4|
|      Albert Collins|        4|
|       Johnny Winter|        4|
+--------------------+---------+
only showing top 20 rows



c)How many songs are included in every Genre?

In [18]:
query = """
        SELECT Genre, COUNT(*) AS num_songs
        FROM songs
        GROUP BY Genre
        """
result = spark.sql(query)

result.show()

+-----+---------+
|Genre|num_songs|
+-----+---------+
|    0|      571|
|    7|      462|
|    6|     1879|
|    9|     1652|
|    5|     1175|
|    1|      883|
|   10|     3784|
|    3|      371|
|    8|     1597|
|    2|     1135|
|    4|      372|
+-----+---------+



d)Which artists dominated the charts?

In [19]:
query = """
        SELECT `Artist Name`, COUNT(*) AS num_songs
        FROM songs
        GROUP BY `Artist Name`
        ORDER BY num_songs DESC
        """

result = spark.sql(query)

result.show()

+--------------------+---------+
|         Artist Name|num_songs|
+--------------------+---------+
|     Backstreet Boys|       65|
|            Westlife|       54|
|      Britney Spears|       52|
|  The Rolling Stones|       31|
|                  U2|       27|
|           Metallica|       25|
|     Lata Mangeshkar|       25|
|       Mohammed Rafi|       21|
|         The Beatles|       21|
|       Fleetwood Mac|       20|
|       Kishore Kumar|       20|
|             Nirvana|       19|
|               AC/DC|       19|
|            Coldplay|       18|
|        Led Zeppelin|       18|
|      The Black Keys|       17|
|           Pearl Jam|       17|
|              Mukesh|       16|
|Red Hot Chili Pep...|       15|
|Creedence Clearwa...|       15|
+--------------------+---------+
only showing top 20 rows



e)Recommend at least 5 fun/not-boring songs that can be played at a party, you can use features like energy, danceability etc.. to represent cheerfulness

In [20]:
query = """
        SELECT `Artist Name`, `Track Name`, Popularity, energy, danceability, valence
        FROM songs
        WHERE energy > 0.8 AND danceability > 0.8 AND valence > 0.8
        ORDER BY (energy + danceability + valence + Popularity) DESC
        LIMIT 5
        """

result = spark.sql(query)

result.show()

+--------------------+--------------------+----------+------+------------+-------+
|         Artist Name|          Track Name|Popularity|energy|danceability|valence|
+--------------------+--------------------+----------+------+------------+-------+
|               Riton|Friday (feat. Muf...|      94.0| 0.862|       0.824|  0.801|
|Riton, Nightcrawl...|Friday (feat. Muf...|      92.0| 0.862|       0.824|  0.801|
|       Nelly Furtado|         Promiscuous|      82.0|  0.97|       0.808|  0.868|
|       Isaiah Rashad|Lay Wit Ya (feat....|      73.0| 0.859|       0.925|  0.967|
|      Yung Baby Tate|I Am (feat. Flo M...|      73.0| 0.835|       0.932|  0.926|
+--------------------+--------------------+----------+------+------------+-------+



In [ ]:
#Which genre has the highest average popularity?
spark.sql("SELECT Genre, AVG(Popularity) AS average_popularity FROM spot GROUP BY Genre ORDER BY average_popularity DESC LIMIT 1;").show()
#Display which artists have recorded the most number of songs with a duration of more than 5 minutes
spark.sql("SELECT Artist_Name, COUNT(*) AS song_sum FROM spot WHERE duration_in_min_ms > 300000 GROUP BY Artist_Name ORDER BY song_sum DESC;").show()
#how many songs in every genre

spark.sql("SELECT Genre, COUNT(*) AS song_sum FROM spot GROUP BY Genre;").show()
#which artists dominated the charts
spark.sql("SELECT Artist_Name, COUNT(*) AS song_sum FROM spot GROUP BY Artist_Name ORDER BY song_sum DESC LIMIT 10; ").show()
#Recommend at least 5 fun/not-boring songs that can be played at a party, you can use features like energy, danceability etc.. to represent cheerfulness
spark.sql("SELECT Artist_Name, Track_Name FROM spot WHERE danceability > 0.7 AND energy > 0.7 AND valence > 0.7 AND acousticness < 0.3 AND instrumentalness < 0.3 ORDER BY Popularity DESC LIMIT 5;").show()

# DataFrames

a)
Which genre has the highest average popularity?

In [21]:
average_popularity_per_genre = spark_df.groupBy('Genre').avg('Popularity')
sorted_popularity_per_genre = average_popularity_per_genre.orderBy('avg(Popularity)', ascending=False)
highest_popularity_genre = sorted_popularity_per_genre.first()

print("Genre with the highest average popularity:", highest_popularity_genre['Genre'])

Genre with the highest average popularity: 4


b)
Display which artists have recorded the most number of songs with a duration of more than 5 minutes

In [22]:
from pyspark.sql.functions import col
songs_duration_gt_5_min = spark_df.filter(col('duration_in min/ms') > 5*60000)
artist_song_counts = songs_duration_gt_5_min.groupBy('Artist Name').count()
sorted_artist_song_counts = artist_song_counts.orderBy('count', ascending=False)
print("Artist(s) with the most number of songs with duration > 5 minutes:")
sorted_artist_song_counts.show()

Artist(s) with the most number of songs with duration > 5 minutes:
+--------------------+-----+
|         Artist Name|count|
+--------------------+-----+
|           Metallica|   19|
|                TOOL|   11|
|        Led Zeppelin|   11|
|         Arcade Fire|    8|
|           Pearl Jam|    8|
|           Aerosmith|    8|
|       Joe Bonamassa|    6|
|        Dire Straits|    5|
|                  U2|    5|
|   Avenged Sevenfold|    5|
|Kenny Wayne Shepherd|    5|
|             Pantera|    5|
|               Opeth|    5|
|     The Stone Roses|    4|
|                 Yes|    4|
|      Phani Narayana|    4|
|       The Beta Band|    4|
|Creedence Clearwa...|    4|
|      Albert Collins|    4|
|       Johnny Winter|    4|
+--------------------+-----+
only showing top 20 rows



c)
How many songs are included in every Genre?

In [23]:
songs_per_genre = spark_df.groupBy('Genre').count()
songs_per_genre.show()

+-----+-----+
|Genre|count|
+-----+-----+
|    0|  571|
|    7|  462|
|    6| 1879|
|    9| 1652|
|    5| 1175|
|    1|  883|
|   10| 3784|
|    3|  371|
|    8| 1597|
|    2| 1135|
|    4|  372|
+-----+-----+



d)
Which artists dominated the charts?

In [24]:
total_songs_per_artist = spark_df.groupBy('Artist Name').count()
sorted_songs_per_artist = total_songs_per_artist.orderBy('count', ascending=False)
print("Top Artists Dominating the Charts by Number of Songs:")
sorted_songs_per_artist.show()

Top Artists Dominating the Charts by Number of Songs:
+--------------------+-----+
|         Artist Name|count|
+--------------------+-----+
|     Backstreet Boys|   65|
|            Westlife|   54|
|      Britney Spears|   52|
|  The Rolling Stones|   31|
|                  U2|   27|
|           Metallica|   25|
|     Lata Mangeshkar|   25|
|       Mohammed Rafi|   21|
|         The Beatles|   21|
|       Fleetwood Mac|   20|
|       Kishore Kumar|   20|
|             Nirvana|   19|
|               AC/DC|   19|
|            Coldplay|   18|
|        Led Zeppelin|   18|
|      The Black Keys|   17|
|           Pearl Jam|   17|
|              Mukesh|   16|
|Red Hot Chili Pep...|   15|
|Creedence Clearwa...|   15|
+--------------------+-----+
only showing top 20 rows



e)
Recommend at least 5 fun/not-boring songs that can be played at a party, you can use features like energy, danceability etc.. to represent cheerfulness.

In [25]:
fun_songs = spark_df.filter((col('energy') >= 0.8) & (col('danceability') >= 0.8) & (col('valence') >= 0.8))
top_fun_songs = fun_songs.select('Artist Name', 'Track Name', 'Popularity', 'energy', 'danceability', 'valence').orderBy('Popularity', ascending=False).limit(5)
print("Recommended Fun/Not-Boring Songs for a Party:")
top_fun_songs.show(truncate=False)

Recommended Fun/Not-Boring Songs for a Party:
+------------------------------------------------+--------------------------------------------------+----------+------+------------+-------+
|Artist Name                                     |Track Name                                        |Popularity|energy|danceability|valence|
+------------------------------------------------+--------------------------------------------------+----------+------+------------+-------+
|Riton                                           |Friday (feat. Mufasa & Hypeman) - Dopamine Re-Edit|94.0      |0.862 |0.824       |0.801  |
|Riton, Nightcrawlers, Mufasa & Hypeman, Dopamine|Friday (feat. Mufasa & Hypeman) - Dopamine Re-Edit|92.0      |0.862 |0.824       |0.801  |
|Nelly Furtado                                   |Promiscuous                                       |82.0      |0.97  |0.808       |0.868  |
|Yung Baby Tate                                  |I Am (feat. Flo Milli)                            |73.0   

In [ ]:
#a) Which genre has the highest average popularity?
bebe = sdf.groupBy("Genre").agg(avg("Popularity").alias("avg_popularity")).orderBy(desc("avg_popularity"))
brbr = bebe.first()["Genre"]
print(brbr)
#b) Display which artists have recorded the most number of songs with a duration of more than 5 minutes
fivemin = sdf.filter(sdf["duration_in_min_ms"] > 300000).groupBy("Artist_Name").agg(count("*").alias("song_sum")).orderBy(desc("song_sum"))
fivemin.show()

#c) How many songs are included in every Genre?
sese = sdf.groupBy("Genre").agg(count("*").alias("song_sum"))
sese.show()

#d) Which artists dominated the charts?
ar = sdf.groupBy("Artist_Name").agg(count("*").alias("song_sum")).orderBy(desc("song_sum"))
ar.show(10)

#e) Recommend at least 5 fun/not-boring songs
fun = sdf.filter((sdf["danceability"] > 0.7) & (sdf["energy"] > 0.7) & (sdf["valence"] > 0.7) & (sdf["acousticness"] < 0.3) & (sdf["instrumentalness"] < 0.3)).orderBy(desc("Popularity")).select("Artist_Name", "Track_Name").limit(5)
fun.show()

# ML


In [28]:
#importing libraries
from pyspark.ml.feature import VectorAssembler
from pyspark.ml import Pipeline
#our columns that we will use to predict the genre
feature_columns = ['Popularity', 'danceability', 'energy', 'key', 'loudness', 'mode',
                   'speechiness', 'acousticness', 'instrumentalness', 'liveness',
                   'valence', 'tempo', 'duration_in min/ms', 'time_signature']
#making them into 1 column
assembler = VectorAssembler(inputCols=feature_columns, outputCol='features')
#splitting data 80% training and 20 % testing
(training_data, testing_data) = spark_df.randomSplit([0.8, 0.2], seed=42)
#making a pipeline and fitting it on our training data
pipeline = Pipeline(stages=[assembler])
pipeline_model = pipeline.fit(training_data)
#transofrming training&testing data to be able to apply models on it
transformed_training_data = pipeline_model.transform(training_data)
transformed_testing_data = pipeline_model.transform(testing_data)

In [29]:
#importing libraries for classifiers
from pyspark.ml.classification import RandomForestClassifier, LogisticRegression, DecisionTreeClassifier
from pyspark.ml.evaluation import MulticlassClassificationEvaluator
#models
rf = RandomForestClassifier(labelCol="Genre", featuresCol="features", numTrees=10)
lr = LogisticRegression(labelCol="Genre", featuresCol="features", maxIter=10)
dt = DecisionTreeClassifier(labelCol="Genre", featuresCol="features")
#fitting models
rf_model = rf.fit(transformed_training_data)
lr_model = lr.fit(transformed_training_data)
dt_model = dt.fit(transformed_training_data)

In [31]:
#geeting predicitons using our models
rf_predictions = rf_model.transform(transformed_testing_data)
lr_predictions = lr_model.transform(transformed_testing_data)
dt_predictions = dt_model.transform(transformed_testing_data)
#defining evaluator that we will use to evaluate each model
evaluator = MulticlassClassificationEvaluator(labelCol="Genre", predictionCol="prediction", metricName="accuracy")
#evaluating models
rf_accuracy = evaluator.evaluate(rf_predictions)
lr_accuracy = evaluator.evaluate(lr_predictions)
dt_accuracy = evaluator.evaluate(dt_predictions)
#printing accuracy for each model to compare them
print("Random Forest Accuracy:", rf_accuracy)
print("Logistic Regression Accuracy:", lr_accuracy)
print("Decision Tree Accuracy:", dt_accuracy)

Random Forest Accuracy: 0.4942357753811826
Logistic Regression Accuracy: 0.5228709557456304
Decision Tree Accuracy: 0.4737820751208628


Based on the accuracies, the Logistic Regression classifier achieved the highest accuracy of approximately 52.29%. The Random Forest and Decision Tree classifiers achieved slightly lower accuracies of approximately 49.42% and 47.38%, respectively.
So based on these results, the Logistic Regression classifier appears to be the best-performing classifier in this case.